### PR TITLE
Initial refactor of InferCtxt towards unified type checking state and per item side tables

### PR DESCRIPTION
**Pull request by jroesch**
This branch begins the work of unifying our type checking contexts into a single piece of state. The goal is to eventually have a single context that we can pass around instead of the fractured situation we currently have. There are still several things that must be done before beginning to make tables item local:
- [ ] move FulfillmentContext into InferCtxt
- [ ] modify SelectionContext to only take a single context argument
- [ ] remove remaining typer impls 
- [ ] remove the ClosureTyper + Typer trait
- [ ] do some renaming to make these things more applicable to their new roles

r? @nikomatsakis 

As a side note there are a couple oddities that are temporary refactors that will be quickly cleaned up in a follow-up PR.

cc @eddyb @Aatch @arielb1 @nrc

### CODE REVIEW

**Comment by eddyb** [comment]
Placing `Tables` inside `InferCtxt` is better than what I was considering.
We can then have all the nested inference-only type context be managed by `InferCtxt`, which would include write-back/lifting `Tables<'local>` to `Tables<'global>` by resolving all inference variables.

**Comment by nikomatsakis** [comment]
r+ from me once we resolve nits, get things building

File: src/librustc/middle/ty.rs [hunk 1]
```diff
@@ -3195,6 +3222,84 @@ impl<'tcx> CommonTypes<'tcx> {
...
+        transmute_restrictions: RefCell::new(Vec::new()),
+        stability: RefCell::new(stability),
+        selection_cache: traits::SelectionCache::new(),
+        repr_hint_cache: RefCell::new(DefIdMap()),
+        const_qualif_map: RefCell::new(NodeMap()),
+        custom_coerce_unsized_kinds: RefCell::new(DefIdMap()),
+        cast_kinds: RefCell::new(NodeMap()),
+   }, f)
+}
+
```
Comment (eddyb) [inline src/librustc/middle/ty.rs]: This looks like a rebase mishap.

**Pull request merged**